adds new comprehensive IT

keith-turner opened the pull request:
Adds a new integration test that excecise most table operations and most read/write operations.

fixes #3826

keith-turner commented:
Seeing this test in around 30 seconds.

```
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.apache.accumulo.test.ComprehensiveIT
[INFO] Tests run: 8, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 32.84 s -- in org.apache.accumulo.test.ComprehensiveIT
```

ddanielr commented:
It took about 50 seconds on my machine. 

```
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.apache.accumulo.test.ComprehensiveIT
[INFO] Tests run: 8, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 49.98 s -- in org.apache.accumulo.test.ComprehensiveIT
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 8, Failures: 0, Errors: 0, Skipped: 0
```

Running `time mvn --color=always clean verify package -Dtest=foo -Dit.test=org.apache.accumulo.test.ComprehensiveIT` 
completes in about 8 minutes.

However, I'm getting inconsistent failures for the `ComprehensiveIT.testTableProperties` section.
Based on the error states, it looks like properties aren't returning their updated states immediately.  

The TableOperations interface mentions that the getTableProperties method is `asynchronous and eventually consistent` so it's behavior seemed correct. We might need to add some wait/retry logic for the expected states.

keith-turner commented:
> We might need to add some wait/retry logic for the expected states.

I will  make testTableProps retry.  Also remembered I had left a TODO about a different issue w/ eventual consistency in the test, need to address that also.

dlmarion commented:
Do you have thoughts on what we should do with the existing tests that are marked Sunny?

keith-turner commented:
> Do you have thoughts on what we should do with the existing tests that are marked Sunny?

Need to go through them and see if they can be untagged if they are covering similar things to this.  May also find things we can add to this when reviewing.  I can take a pass at that and either open a follow on issue or make the changes in this PR.

dlmarion commented:
For `main`, I think it might make sense to untag duplicate ITs in this PR. When merging to `elasticity`, would need to re-evaluate as we tagged new tests with Sunny. Also for `elasticity`, the scope of read/write tests have changed.

keith-turner commented:
> When merging to elasticity, would need to re-evaluate as we tagged new tests with Sunny. Also for elasticity, the scope of read/write tests have changed.

Also opened #3834 for elasticity, if something is done for that it will change sunny in elasticity.

keith-turner commented:
I reviewed the ITs tagged for sunny and kept the following. 

 * InMemoryMapIT : because its super fast
 * NativeMapIT :  because its super fast
 * ReadWriteIT : because it test monitor and cluster shutdown
 * WALSunnyDayIT : because it test log recovery and has sunny in name
 * ThriftBehaviorIT :  because its super fast
 * KeywordStartIT :  because its super fast
  
Removed the sunny tag from the following.
  * PropStoreConfigIT : have basic coverage of setting per table props which exercises the prop store, also added tests setting namespace props
  * ZooPropEditorIT : this seems like a test for a more obscure feature, seems ok to leave it for a full IT run that would be done prior to release
  * BulkIT ... have coverage of bulk import
  *  ScanIteratorIT : improved iterator test coverage in ComprehensiveIT  .. also plan to add test that creates user with auths
  
Still looking at the Shell ITs.  Think if some basic tests of users and table permissions were added to ComprehensiveIT that these could be removed from sunny tag.

dlmarion commented on test/src/main/java/org/apache/accumulo/test/ComprehensiveIT.java:
```diff
@@ -0,0 +1,946 @@
+  public void testBulkImport() throws Exception {
+    String table = getUniqueNames(1)[0];
+    try (AccumuloClient client = Accumulo.newClient().from(getClientProps()).build()) {
+      client.tableOperations().create(table);
```
On the merge to `elasticity`, you'll likely want to use the create table with options and set the hosting goal to always for all of the tables created in this test.

keith-turner replied:
@dlmarion I will merge these changes to elasticity after I make the follow on PR to add security.  I also look at getting the test to run then.

ddanielr commented:
Wait changes tested again. 
Average IT completion time is now ~60 seconds and the total runtime of `time mvn --color=always clean verify package -Dtest=foo -Dit.test=org.apache.accumulo.test.ComprehensiveIT` is around 7.5 min long.

keith-turner commented:
I am going to go ahead an merge this since its been approved so many times.  I plan to shortly open another PR to add one more test that covers some basic security operations and remove shell ITs from sunny.  That addition can be looked at in isolation.
